compiler/rustc_mir_build/src/build/scope.rs

```diff
@@ -1141,6 +1142,9 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     {
                         return;
                     }
+                    // Opaque type may not have been scheduled if its underlying
+                    // type does not need drop.
+                    None if self.local_decls[local].ty.has_opaque_types() => return,
```

Comment:
that doesn't seem right, though I am still not fully comfortable with this code. Logging both `schedule_drop` and the scope of the `bug!` results in 
```
lcnr@lcnr-PC:~/test1$ rustc +rust2 src/main.rs
  WARN rustc_mir_build::build::scope considering dropping bool in scope Arguments(14)
  WARN rustc_mir_build::build::scope not dropping bool
  WARN rustc_mir_build::build::scope considering dropping bool in scope Node(3)
  WARN rustc_mir_build::build::scope considering dropping bool in scope Node(3)
  WARN rustc_mir_build::build::scope not dropping bool
  WARN rustc_mir_build::build::scope considering dropping Alias(Opaque, AliasTy { args: [], def_id: DefId(0:8 ~ main[7eb4]::if_else::{opaque#0}) }) in scope CallSite(14)
  WARN rustc_mir_build::build::scope considering dropping Alias(Opaque, AliasTy { args: [], def_id: DefId(0:8 ~ main[7eb4]::if_else::{opaque#0}) }) in scope CallSite(14)
  WARN rustc_mir_build::build::scope considering dropping bool in scope Arguments(10)
  WARN rustc_mir_build::build::scope not dropping bool
  WARN rustc_mir_build::build::scope considering dropping bool in scope Node(3)
  WARN rustc_mir_build::build::scope considering dropping bool in scope Node(3)
  WARN rustc_mir_build::build::scope not dropping bool
error: internal compiler error: compiler/rustc_mir_build/src/build/scope.rs:1149:26: found wrong drop, expected value drop of _0 in scope CallSite(10), found None
```
so we're adding the opaque at call site `14` but drop it at call site `10`?

Reply:
so the affected code is
```rust
fn if_no_else(c: bool) -> impl Sized {
    if c {}
}
```
where the opaque type gets unified with `()` and `()` does not need drop while `impl Sized` does? :thinking: we probably never emit drop for the return value of an `if` without an else block or sth?